packages/components/utils/src/useKeyboard/useKeyboard.ts

```diff
@@ -0,0 +1,60 @@
+import { useEffect, useCallback, MutableRefObject } from 'react';
+
+export interface UseKeyboardProps {
+  /**
+   * @description object of key names and handlers defines which key to look for i.e. `ArrowUp`, `Escape`, `Shift`
+   * value is a callback function to be called when key matches
+   */
+  keys: {
+    [key: string]: (e: KeyboardEvent) => void;
+  };
+  /**
+   * @description defines the attached event type
+   * @default 'keydown'
+   */
+  event?: 'keyup' | 'keypress' | 'keydown';
+
+  /**
+   * @description React reference to attach the event to its current element
+   */
+  ref?: MutableRefObject<HTMLElement>;
+}
+
+/**
+ *
+ * @description hook to attach a handler keyboard event listener to `document` or `HTMLElements` with garbage collection
+ * @example
+ * useKeyboard({
+ *  event: 'keydown', // Optional, default is `keydown`
+ *  ref: yourReactRef, // Optional, by default event is attached to document
+ *  keys: {
+ *    ArrowUp: (e) => handleArrowUp(e),
+ *    Tab: (e) => handleTab(e)
+ *  }
+ * })
+ */
+export const useKeyboard = (props: UseKeyboardProps) => {
+  const { ref, keys, event = 'keydown' } = props;
+  let element: HTMLElement | Document = document;
+
+  const handleKeyEvent = useCallback(
+    (e) => {
+      const isKey = Object.prototype.hasOwnProperty.call(keys, e.key);
+      if (isKey) {
+        keys[e.key](e);
+      }
+    },
+    [keys],
+  );
+
+  useEffect(() => {
+    if (ref) {
+      element = ref.current;
+    }
+
+    element.addEventListener(event, handleKeyEvent);
+    return () => {
+      element.removeEventListener(event, handleKeyEvent);
+    };
+  }, [element, ref, event, handleKeyEvent]);
```

Comment:
Hey you're right, element is not necessary as it doesn't change, left over from previous implementation.
`ref` is there because as it should be watched for changes and re-trigger the events addition/removal.

Reply:
But `ref` is a mutable object, it will not change. So when ref.current changes its value from `null` to some `element` it won't trigger this effect.

Reply:
The goal here is not to monitor `ref.current`, it's to monitor `ref` itself as it's `undefined` by default if the consumer doesn't pass it we fallback to `document`

Reply:
I removed it and the events gets attached properly. also I don't think users will dynamically change `ref` so no need to watch it as well. Thanks for the feedback 👍🏼

Reply:
I don't agree that users will not dynamically change ref, here is the example: https://github.com/contentful/forma-36/pull/1343 
In this case I can't attach events on the button with ref. It attaches to document, since the ref.current is `null` on the first render.

Reply:
and what if I don't want to attach the events to a document when my `ref.current` is `null`? Also something to consider

Reply:
that shouldn't be the case. If you provide a ref it must be attached to the current. I'll have a look, thanks